packages/animations/lib/src/dual_transition_builder.dart

```diff
@@ -0,0 +1,228 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+
+/// Builder callback used by [DualTransitionBuilder].
+///
+/// The builder is expected to return a transition powered by the provided
+/// `animation` and wrapping the provided `child`.
+///
+/// The `animation` provided to the builder always runs forward from 0.0 to 1.0.
+/// The builder will be called whenever the [Animation.value] of `animation`
+/// has changed.
+typedef TransitionBuilder = Widget Function(
+  BuildContext context,
+  Animation<double> animation,
+  Widget child,
+);
+
+/// A animated builder that animates its [child] with different transitions
+/// based on the [AnimationStatus] of the provided [animation].
+///
+/// While the [animation] runs forward, the [child] is animated according to
+/// [forwardBuilder] and while the [animation] is running in reverse, it is
+/// animated according to [reverseBuilder].
+///
+/// This widget can be used to specify different enter and exit transitions for
+/// a [child].
```

Comment:
Maybe move this up to the second paragraph, since this provides higher level context on the purpose of this widget, while the paragraph prior to this one is more granular detail.